File: src/main/java/cpsc224/creatures/CreatureFactory.java
package cpsc224.creatures;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.ArrayList;
import java.util.List;

import cpsc224.items.Inventory;
import cpsc224.items.consumables.ConsumableFactory;
import cpsc224.items.weapons.WeaponFactory;
import cpsc224.utils.BufferedImageBuilder;

import javax.swing.*;

/**
 * A factory class to create creatures.
 */

public class CreatureFactory {

    private static ImageIcon getIcon(String name, boolean flipHorizontally) {
        BufferedImageBuilder imageBuilder = new BufferedImageBuilder("/images/creatures/" + name + ".png");

        if (flipHorizontally)
            imageBuilder.flipHorizontally();

        return imageBuilder
                .scale(256, 256)
                .toImageIcon();
    }

    private static ImageIcon getIcon(String name) {
        return getIcon(name, true);
    }


    public static Player createPlayer() {
        final String name = "Player";
        final int health = 200;
        final CreatureModifiers cm = new CreatureModifiers(1, .2, new LinkedList<>(Arrays.asList(1.0, 1.0, 1.0, 1.0)));
        final Inventory inv = new Inventory(4, 4);
        inv.setWeapon(0, WeaponFactory.createRustyDagger());
        inv.setConsumable(0, ConsumableFactory.createSmallHealthPotion());
        inv.setConsumable(1, ConsumableFactory.createWeaponRefillPotion());

        return new Player(name, health, cm, inv);
    }

    public static Player createMagePlayer() {
        final String name = "Mage";
        final int health = 200;
        final CreatureModifiers cm = new CreatureModifiers(1, .2, new LinkedList<>(Arrays.asList(1.0, 1.0, 1.0, 1.0)));
        final Inventory inv = new Inventory(4, 4);
        inv.setWeapon(0, WeaponFactory.createBasicWand());
        inv.setConsumable(0, ConsumableFactory.createSmallHealthPotion());
        inv.setConsumable(1, ConsumableFactory.createWeaponRefillPotion());


        return new Player(name, health, cm, inv);
    }

    public static Player createBluntPlayer() {
        final String name = "Barbarian";
        final int health = 200;
        final CreatureModifiers cm = new CreatureModifiers(1, .2, new LinkedList<>(Arrays.asList(1.0, 1.0, 1.0, 1.0)));
        final Inventory inv = new Inventory(4, 4);
        inv.setWeapon(0, WeaponFactory.createWoodClub());
        inv.setConsumable(0, ConsumableFactory.createSmallHealthPotion());
        inv.setConsumable(1, ConsumableFactory.createWeaponRefillPotion());

        return new Player(name, health, cm, inv);
    }

    public static Player createSlashPlayer() {
        final String name = "Knight";
        final int health = 200;
        final CreatureModifiers cm = new CreatureModifiers(1, .2, new LinkedList<>(Arrays.asList(1.0, 1.0, 1.0, 1.0)));
        final Inventory inv = new Inventory(4, 4);
        inv.setWeapon(0, WeaponFactory.createRustyDagger());
        inv.setConsumable(0, ConsumableFactory.createSmallHealthPotion());
        inv.setConsumable(1, ConsumableFactory.createWeaponRefillPotion());

        return new Player(name, health, cm, inv);
    }

    public static Player createRangePlayer() {
        final String name = "Ranger";
        final int health = 200;
        final CreatureModifiers cm = new CreatureModifiers(1, .2, new LinkedList<>(Arrays.asList(1.0, 1.0, 1.0, 1.0)));
        final Inventory inv = new Inventory(4, 4);
        inv.setWeapon(0, WeaponFactory.createCrudeBow());
        inv.setConsumable(0, ConsumableFactory.createSmallHealthPotion());
        inv.setConsumable(1, ConsumableFactory.createWeaponRefillPotion());

        return new Player(name, health, cm, inv);
    }


    public static Creature createRat() {
        final String name = "Rat";
        final int health = 10;
        final CreatureModifiers cm = new CreatureModifiers(1, .15, new LinkedList<>(Arrays.asList(1.0, 1.0, 1.0, 1.0)));
        final Inventory inv = new Inventory(4, 0);
        inv.setWeapon(0, WeaponFactory.createRatClaws());
        List<Double> weaponWeights = new ArrayList<Double>();
        weaponWeights.add(1.0);

        return new Creature(name, health, cm, inv, getIcon(name), weaponWeights);
    }

    public static Creature createTroll() {
        final String name = "Troll";
        final int health = 150;
        final CreatureModifiers cm = new CreatureModifiers(2, .05, new LinkedList<>(Arrays.asList(0.7, 1.5, 0.8, 2.0)));
        final Inventory inv = new Inventory(4, 0);
        inv.setWeapon(0, WeaponFactory.createWoodClub());
        List<Double> weaponWeights = new ArrayList<Double>();
        weaponWeights.add(1.0);

        return new Creature(name, health, cm, inv, weaponWeights);
    }

    public static Creature createBird() {
        final String name = "Bird";
        final int health = 30;
        final CreatureModifiers cm = new CreatureModifiers(3, .2, new LinkedList<>(Arrays.asList(0.5, 0.7, 2.0, 1.4)));
        final Inventory inv = new Inventory(4,0);
        inv.setWeapon(0,WeaponFactory.createBirdTalons());
        List<Double> weaponWeights = new ArrayList<Double>();
        weaponWeights.add(1.0);

        return new Creature(name, health, cm, inv, weaponWeights);
    }

    public static Creature createWitch() {
        final String name = "Witch";
        final int health = 75;
        final CreatureModifiers cm = new CreatureModifiers(2, .1, new LinkedList<>(Arrays.asList(1.2, 1.4, 0.8, 0.7)));
        final Inventory inv = new Inventory(4,0);
        inv.setWeapon(0, WeaponFactory.createToxicStaff());
        inv.setWeapon(1, WeaponFactory.createSteelSword());
        List<Double> weaponWeights = new ArrayList<Double>();
        weaponWeights.add(4.0);
        weaponWeights.add(1.0);

        return new Creature(name, health, cm, inv, weaponWeights);
    }

    public static Creature createGuardian() {
        final String name = "Guardian";
        final int health = 105;
        final CreatureModifiers cm = new CreatureModifiers(2, .15, new LinkedList<>(Arrays.asList(0.5, 0.8, 1.1, 2.0)));
        final Inventory inv = new Inventory(4,0);
        inv.setWeapon(0, WeaponFactory.createRoyalSword());
        inv.setWeapon(1, WeaponFactory.createSteelHammer());
        List<Double> weaponWeights = new ArrayList<Double>();
        weaponWeights.add(2.0);
        weaponWeights.add(1.0);
        
        return new Creature(name, health, cm, inv, weaponWeights);
    }

    public static Creature createGoblin() {
        final String name = "Goblin";
        final int health = 50;
        final CreatureModifiers cm = new CreatureModifiers(1, 0.1, new LinkedList<>(Arrays.asList(1.0, 0.9, 1.2, 1.4)));
        final Inventory inv = new Inventory(4,0);
        inv.setWeapon(0, WeaponFactory.createRustyDagger());
        inv.setWeapon(1, WeaponFactory.createWoodClub());
        inv.setWeapon(2, WeaponFactory.createSteelSword());
        List<Double> weaponWeights = new ArrayList<Double>();
        weaponWeights.add(1.5);
        weaponWeights.add(1.0);
        weaponWeights.add(0.5);

        return new Creature(name, health, cm, inv, getIcon(name), weaponWeights);
    }

    public static Creature createSkeleton() {
        final String name = "Skeleton";
        final int health = 50;
        final CreatureModifiers cm = new CreatureModifiers(1, 0.1, new LinkedList<>(Arrays.asList(1.5, 0.5, 0.5, 2.0)));
        final Inventory inv = new Inventory(4,0);
        inv.setWeapon(0, WeaponFactory.createCrossbow());
        inv.setWeapon(1, WeaponFactory.createRustyDagger());
        List<Double> weaponWeights = new ArrayList<Double>();
        weaponWeights.add(3.0);
        weaponWeights.add(1.0);

        return new Creature(name, health, cm, inv, weaponWeights);
    }


    // Bosses

    public static Creature createFireGolem() {
        final String name = "Fire golem";
        final int health = 500;
        final CreatureModifiers cm = new CreatureModifiers(1.5, 0.1, new LinkedList<>(Arrays.asList(1.0, 1.0, 1.5, 0.5)));
        final Inventory inv = new Inventory(4,0);
        inv.setWeapon(0, WeaponFactory.creatFireHeal());
        inv.setWeapon(1, WeaponFactory.createFireBall());
        inv.setWeapon(2, WeaponFactory.createFireSword());
        inv.setWeapon(3, WeaponFactory.createFireClaws());
        List<Double> weaponWeights = new ArrayList<Double>();

        weaponWeights.add(1.0);
        weaponWeights.add(2.0);
        weaponWeights.add(2.0);
        weaponWeights.add(1.5);

        return new Creature(name, health, cm, inv, getIcon("FireGolemBoss"),weaponWeights);
    }

    public static Creature createRockGolem() {
        final String name = "Rock golem";
        final int health = 800;
        final CreatureModifiers cm = new CreatureModifiers(1, 0.1, new LinkedList<>(Arrays.asList(0.5,0.5,0.5,2.0)));
        final Inventory inv = new Inventory(4,0);
        inv.setWeapon(0, WeaponFactory.createRockHeal());
        inv.setWeapon(1, WeaponFactory.createRockSlam());
        inv.setWeapon(2, WeaponFactory.createRockThrow());
        inv.setWeapon(3, WeaponFactory.createRockCut());
        List<Double> weaponWeights = new ArrayList<Double>();


        weaponWeights.add(1.0);
        weaponWeights.add(2.0);
        weaponWeights.add(2.0);
        weaponWeights.add(1.5);

        return new Creature(name, health, cm, inv, getIcon("RockGolemBoss"), weaponWeights);
    }

    public static Creature createIceGolem() {
        final String name = "Ice golem";
        final int health = 500;
        final CreatureModifiers cm = new CreatureModifiers(1.5, 0.1, new LinkedList<>(Arrays.asList(1.5, 1.5, 1.0, 0.5)));
        final Inventory inv = new Inventory(4,0);
        inv.setWeapon(0, WeaponFactory.creatFrostHeal());
        inv.setWeapon(1, WeaponFactory.createIceball());
        inv.setWeapon(2, WeaponFactory.createIceSpear());
        inv.setWeapon(3, WeaponFactory.createIceClaws());
        List<Double> weaponWeights = new ArrayList<Double>();

        weaponWeights.add(1.0);
        weaponWeights.add(2.0);
        weaponWeights.add(2.0);
        weaponWeights.add(1.5);

        return new Creature(name, health, cm, inv, getIcon("IceGolemBoss"), weaponWeights);
    }
}

File: web/src/game/creatures/CreatureFactory.ts
import { CreatureModifiers } from './CreatureModifiers';
import { Creature } from './Creature';
import { Player } from './Player';
import { Inventory } from '../items/Inventory';
import { WeaponFactory } from '../items/weapons/WeaponFactory';
import { ConsumableFactory } from '../items/consumables/ConsumableFactory';
import { creatureIcon } from '../utils/icons';

function mods(damage: number, evasion: number, resists: number[]): CreatureModifiers {
  return new CreatureModifiers(damage, evasion, resists);
}

export const CreatureFactory = {
  createPlayer(): Player {
    const inv = new Inventory(4, 4, 3);
    inv.addItem(WeaponFactory.createRustyDagger());
    inv.addItem(ConsumableFactory.createSmallHealthPotion());
    inv.addItem(ConsumableFactory.createWeaponRefillPotion());
    return new Player('Player', 200, mods(1, 0.2, [1, 1, 1, 1]), inv, null);
  },

  createMagePlayer(): Player {
    const inv = new Inventory(4, 4, 3);
    inv.addItem(WeaponFactory.createEnchantedStaff());
    inv.addItem(WeaponFactory.createRustyDagger());
    inv.addItem(ConsumableFactory.createSmallHealthPotion());
    inv.addItem(ConsumableFactory.createWeaponRefillPotion());
    return new Player('Mage', 200, mods(1, 0.2, [1, 1, 1, 0.8]), inv, creatureIcon('mage2', false));
  },

  createBluntPlayer(): Player {
    const inv = new Inventory(4, 4, 3);
    inv.addItem(WeaponFactory.createSteelMace());
    inv.addItem(WeaponFactory.createCrudeBow());
    inv.addItem(ConsumableFactory.createSmallHealthPotion());
    inv.addItem(ConsumableFactory.createWeaponRefillPotion());
    return new Player('Barbarian', 200, mods(1, 0.2, [0.8, 1, 1, 1]), inv, creatureIcon('Barb2', false));
  },

  createSlashPlayer(): Player {
    const inv = new Inventory(4, 4, 3);
    inv.addItem(WeaponFactory.createSteelSword());
    inv.addItem(WeaponFactory.createWoodClub());
    inv.addItem(ConsumableFactory.createSmallHealthPotion());
    inv.addItem(ConsumableFactory.createWeaponRefillPotion());
    return new Player('Knight', 200, mods(1, 0.2, [1, 0.8, 1, 1]), inv, creatureIcon('knight2', false));
  },

  createRangePlayer(): Player {
    const inv = new Inventory(4, 4, 3);
    inv.addItem(WeaponFactory.createLongBow());
    inv.addItem(WeaponFactory.createRudimentaryStaff());
    inv.addItem(ConsumableFactory.createSmallHealthPotion());
    inv.addItem(ConsumableFactory.createWeaponRefillPotion());
    // +15% dodge (evasion 0.35); projectile accuracy bonus applied in Fight
    return new Player('Ranger', 200, mods(1, 0.35, [1, 1, 0.8, 1]), inv, creatureIcon('Ranger', false));
  },

  createRat(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createRatClaws());
    return new Creature('Rat', 10, mods(1, 0.15, [1, 1, 1, 1]), inv, creatureIcon('Rat', true), [1]);
  },

  createSlime(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createRatClaws());
    return new Creature('Slime', 25, mods(0.8, 0.25, [0.8, 1.2, 0.8, 1.0]), inv, creatureIcon('Slime', false), [1]);
  },

  createSorcerer(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createToxicStaff());
    inv.addItem(WeaponFactory.createEnchantedStaff());
    return new Creature('Sorcerer', 75, mods(1.5, 0.1, [1.2, 1.4, 0.8, 0.7]), inv, creatureIcon('Sorcerer', true), [2, 3]);
  },

  createMushroom(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createMushroomCap());
    return new Creature('Mushroom', 40, mods(1, 0.1, [1.2, 1, 1, 0.8]), inv, creatureIcon('Mushroom', false), [1]);
  },

  createGoblin(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createRustyDagger());
    inv.addItem(WeaponFactory.createWoodClub());
    return new Creature('Goblin', 50, mods(1, 0.1, [1, 0.9, 1.2, 1.4]), inv, creatureIcon('Goblin', true), [1.5, 1]);
  },

  createSkeleton(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createLongBow());
    inv.addItem(WeaponFactory.createRustyDagger());
    return new Creature('Skeleton', 50, mods(1, 0.1, [1.5, 0.5, 0.5, 2.0]), inv, creatureIcon('Skeleton', true), [3, 1]);
  },

  createGuardian(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createRoyalSword());
    inv.addItem(WeaponFactory.createSteelHammer());
    return new Creature('Guardian', 105, mods(1.5, 0.1, [0.5, 0.8, 1.1, 2.0]), inv, creatureIcon('PurpleGolemBoss', true), [2, 1]);
  },

  createBat(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createBirdTalons());
    inv.addItem(WeaponFactory.createWebCaster());
    return new Creature('Bat', 28, mods(1.1, 0.35, [1.1, 1, 0.7, 1.2]), inv, creatureIcon('Rat', true), [2, 1]);
  },

  createImp(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createEmberStaff());
    inv.addItem(WeaponFactory.createSparkRod());
    return new Creature('Imp', 45, mods(1.2, 0.2, [1.2, 1.1, 1, 0.7]), inv, creatureIcon('Sorcerer', true), [2, 1]);
  },

  createBandit(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createCleaver());
    inv.addItem(WeaponFactory.createJavelin());
    return new Creature('Bandit', 60, mods(1.15, 0.18, [1, 0.85, 1.1, 1.2]), inv, creatureIcon('Goblin', true), [2, 1.5]);
  },

  createWraith(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createFrostWand());
    inv.addItem(WeaponFactory.createDrainWand());
    return new Creature('Wraith', 70, mods(1.3, 0.25, [1.4, 1.3, 0.9, 0.55]), inv, creatureIcon('Skeleton', true), [2, 1]);
  },

  createBrute(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createStunMace());
    inv.addItem(WeaponFactory.createWoodClub());
    return new Creature('Brute', 90, mods(1.4, 0.05, [0.65, 1.2, 1, 1.5]), inv, creatureIcon('Barb2', false), [2, 1]);
  },

  createCultist(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createToxicStaff());
    inv.addItem(WeaponFactory.createHealStaff());
    return new Creature('Cultist', 80, mods(1.25, 0.12, [1.15, 1.2, 0.85, 0.65]), inv, creatureIcon('mage2', false), [2, 1]);
  },

  createSporekin(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createMushroomCap());
    inv.addItem(WeaponFactory.createVenomFang());
    return new Creature('Sporekin', 55, mods(1, 0.15, [1.1, 1, 1.1, 0.75]), inv, creatureIcon('Mushroom', false), [2, 1]);
  },

  createScout(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createBoneBow());
    inv.addItem(WeaponFactory.createRustyDagger());
    return new Creature('Scout', 48, mods(1.1, 0.3, [1, 1, 0.75, 1.15]), inv, creatureIcon('Ranger', false), [3, 1]);
  },

  createTroll(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createWoodClub());
    inv.addItem(WeaponFactory.createStunMace());
    return new Creature('Troll', 150, mods(2, 0.05, [0.7, 1.5, 0.8, 2.0]), inv, creatureIcon('Barb2', false), [2, 1]);
  },

  createBird(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createBirdTalons());
    inv.addItem(WeaponFactory.createWebCaster());
    return new Creature('Bird', 30, mods(3, 0.2, [0.5, 0.7, 2.0, 1.4]), inv, creatureIcon('Rat', true), [2, 1]);
  },

  /** Floor 1 boss — multi-kit situation, not a golem wall yet. */
  createPurpleGolem(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createStunMace());
    inv.addItem(WeaponFactory.createEmberStaff());
    inv.addItem(WeaponFactory.createCleaver());
    return new Creature(
      'Purple golem',
      140,
      mods(1.3, 0.08, [0.75, 0.85, 1.1, 0.9]),
      inv,
      creatureIcon('PurpleGolemBoss', true),
      [2, 2, 1.5],
    );
  },

  createFireGolem(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.creatFireHeal());
    inv.addItem(WeaponFactory.createFireBall());
    inv.addItem(WeaponFactory.createFireSword());
    inv.addItem(WeaponFactory.createFireClaws());
    return new Creature(
      'Fire golem',
      200,
      mods(1.5, 0.05, [1, 1, 1.5, 0.5]),
      inv,
      creatureIcon('FireGolemBoss', true),
      [1, 2, 2, 1.5],
    );
  },

  createRockGolem(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.createRockHeal());
    inv.addItem(WeaponFactory.createRockSlam());
    inv.addItem(WeaponFactory.createRockThrow());
    inv.addItem(WeaponFactory.createRockCut());
    return new Creature(
      'Rock golem',
      200,
      mods(1, 0.05, [0.5, 0.5, 0.5, 2.0]),
      inv,
      creatureIcon('RockGolemBoss', true),
      [1, 2, 2, 1.5],
    );
  },

  createIceGolem(): Creature {
    const inv = new Inventory(4, 0);
    inv.addItem(WeaponFactory.creatFrostHeal());
    inv.addItem(WeaponFactory.createIceball());
    inv.addItem(WeaponFactory.createIceSpear());
    inv.addItem(WeaponFactory.createIceClaws());
    return new Creature(
      'Ice golem',
      200,
      mods(1.5, 0.05, [1.5, 1.5, 1, 0.5]),
      inv,
      creatureIcon('IceGolemBoss', true),
      [1, 2, 2, 1.5],
    );
  },
};
